Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
++    protected abstract void removePlayerTicketAreaMap(EntityPlayer player);
++    //Paper end
+ 
+     protected ChunkMapDistance(Executor executor, Executor executor1) {
+         executor1.getClass();
+@@ -234,6 +238,7 @@ public abstract class ChunkMapDistance {
+         ((ObjectSet) this.c.computeIfAbsent(i, (j) -> {
+             return new ObjectOpenHashSet();
+         })).add(entityplayer);
++        this.updatePlayerTicketAreaMap(entityplayer, i); //Paper ticket level pp-view-dist
+         this.f.b(i, 0, true);
+         this.g.b(i, 0, true);
+     }
+@@ -251,6 +256,10 @@ public abstract class ChunkMapDistance {
+ 
+     }
+ 
++    public void fullRemove(EntityPlayer player) {
++        this.removePlayerTicketAreaMap(player); //Paper ticket level pp-view-dist
++    }
++
+     protected String c(long i) {
+         ObjectSortedSet<Ticket<?>> objectsortedset = (ObjectSortedSet) this.tickets.get(i);
+         String s;
+@@ -278,6 +287,10 @@ public abstract class ChunkMapDistance {
+         return this.f.a.containsKey(i);
+     }
+ 
++    public boolean naturalSpawnChunksHave(long i) {
++        return this.f.a.containsKey(i); //Paper ticket level pp-view-dist
++    }
++
+     public String c() {
+         return this.i.a();
+     }
+@@ -372,20 +385,20 @@ public abstract class ChunkMapDistance {
+                 byte b0 = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getByteValue();
+                 long j = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getLongKey();
+ 
+-                this.a(j, b0, this.c(b0), b0 <= i - 2);
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i); //Paper ticket level pp-view-dist
+             }
+ 
+             this.e = i;
+         }
+ 
+-        private void a(long i, int j, boolean flag, boolean flag1) {
++        private void a(long i, int j, boolean flag, boolean flag1, int vd) { //Paper ticket level pp-view-dist
+             if (flag != flag1) {
+                 Ticket<?> ticket = new Ticket<>(TicketType.PLAYER, ChunkMapDistance.b, new ChunkCoordIntPair(i), ChunkMapDistance.this.currentTick);
+ 
+                 if (flag1) {
+                     ChunkMapDistance.this.j.a(ChunkTaskQueueSorter.a(() -> { // Craftbukkit - decompile error
+                         ChunkMapDistance.this.m.execute(() -> {
+-                            if (this.c(this.c(i))) {
++                            if (this.c(i) <= vd) { //Paper ticket level pp-view-dist
+                                 ChunkMapDistance.this.addTicket(i, ticket);
+                                 ChunkMapDistance.this.l.add(i);
+                             } else {
+@@ -430,13 +443,46 @@ public abstract class ChunkMapDistance {
+                             }
+ 
+                         });
+-                        this.a(i, k, this.c(j), this.c(k));
+                     }
+-                }
+ 
++                    //Paper start ticket level pp-view-dist
```

Comment:
the distance map utility should have this logic, given it's intentionally capable of replacing mojang's utility
that way we don't need to worry about any changes mojang decides to make in the future about how they propagate updates
obfhelpers on the ChunkTaskQueueSorter should be enough to diff if their logic changes in that class

Reply:
solved in rewrite already